### PR TITLE
[NVIDIA XLA GPU] Fix allreduce/reducescatter combiner bug

### PR DESCRIPTION
**Pull request by Tixxx**
AR/RS combiners are run after fusion and float normalization passes. This will make them fail to combine anything because:
1. float normalization will promote to_apply regions of ARs and RSs by inserting casts.
2. fusion will fuse the cast-compute pattern into a single fusion kernel

Then combiners won't be able to match for collectives of the same type since it doesn't look into the fusions to determine allreduce keys.
This PR adds mechanisms for float normalizer to skip processing to_apply regions of collectives so they are untouched throughout the whole compilation.

### CODE REVIEW

**Comment by jurahul** [comment]
My overall comment is that we have this kind of matching in other places as well (somewhere in XLA/GPU runtime lowering) and the ideal solution here is to leave the bf16 computation attached to all-reduce/reduce-scatter as is during promotion so we don't have to look through this fusion in other places.

**Comment by Tixxx** [comment]
> My overall comment is that we have this kind of matching in other places as well (somewhere in XLA/GPU runtime lowering) and the ideal solution here is to leave the bf16 computation attached to all-reduce/reduce-scatter as is during promotion so we don't have to look through this fusion in other places.

Yea, currently this only happens in float normalization but I feel this type of promotion can happen anywhere and this issue might come back some time in the future. For now I can look into float normalization to prevent it from promoting collectives' called computations. But I'm thinking if there's a more explicit way to do this, something like giving collective instructions another field called unmutable_called_computation which is read-only so their called_computations is empty therefore won't be modified.

**Comment by Tixxx** [comment]
> My overall comment is that we have this kind of matching in other places as well (somewhere in XLA/GPU runtime lowering) and the ideal solution here is to leave the bf16 computation attached to all-reduce/reduce-scatter as is during promotion so we don't have to look through this fusion in other places.

@jurahul  I changed float normalization to not promote to_apply regions of collectives. Let me know if this would make more sense than the first commit. Thanks.

**Review by jurahul — CHANGES_REQUESTED** [review]
I agree that this change can be done first. I think it might be better to split the HLO computation changes and float normalization changes into 2 separate PRs.

File: xla/service/float_normalization_test.cc [hunk 1]
```diff
@@ -485,4 +527,135 @@ TEST_F(FloatNormalizationTest, ResolveIfUnsupportedF8e5m2) {
...
+                         /*allow_mixed_precision=*/true);
+    EXPECT_IS_OK(verifier.Run(module).status());
+
+    return result.value();
+  }
+};
+
+TEST_F(FloatNormalizationNoComputeSupportTest,
+       NoNormalizationForToApplyMultiOuputAllReduce) {
+  auto module = CreateNewVerifiedModule();
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/float_normalization_test.cc:549]: Agreed, i thought about it but all tests in this file are following this pattern(including some of the existing tests that contain collectives) so I didnt want to break the style.

File: xla/service/float_normalization_test.cc [hunk 2]
```diff
@@ -76,6 +76,48 @@ class TestFloatSupport : public FloatSupport {
   }
 };
 
+class TestFloatNoComputeSupport : public FloatSupport {
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/float_normalization_test.cc:79]: the class TestFloatSupport supports some elementwise compute ops to be low precision but doesnt support collectives, that's actually not the case we wanted. We would need a floatSupport class that doesn't support elementwise compute ops but supports collective ops to repro the issue. That's why I created a new class based on `FloatSupport`.
I will remove some redundant code in this class that wont be needed by the use-case of this class and put some comments on top of the class.

**Comment by Tixxx** [comment]
> I agree that this change can be done first. I think it might be better to split the HLO computation changes and float normalization changes into 2 separate PRs.

Sure, I have extracted out the hlo computation changes and created a separate pr, https://github.com/openxla/xla/pull/6171.
Could you help review that? I will rebase this one after the above is merged.

**Review by jurahul — APPROVED** [review]

**Review by frgossen — APPROVED** [review]
Thanks!

**Review by ddunl — APPROVED** [review]
Copybara not picking up on other reviews for some reason...

**Comment by frgossen** [comment]
The breakage is an HLO shape mismatch after float-normalization. I will try to get you more info on this

**Comment by Tixxx** [comment]
> The breakage is an HLO shape mismatch after float-normalization. I will try to get you more info on this

@frgossen Thanks for the heads-up. If you have a snippet of the erroneous hlo, that would be great.

**Pull request closed without merging**